File: library/src/main/java/com/hextremelabs/permiscus/PermissionRequest.java
package com.hextremelabs.permiscus;

import com.hextremelabs.permiscus.callbacks.OnPermissionDeniedCallback;
import com.hextremelabs.permiscus.callbacks.OnPermissionGrantedCallback;
import com.hextremelabs.permiscus.callbacks.OnPermissionShowRationaleCallback;
import androidx.annotation.NonNull;

public class PermissionRequest {
    private final @NonNull PermissionManager permissionManager;

    private final @NonNull String[] permissions;
    private final int requestCode;

    protected final OnPermissionGrantedCallback grantedCallback;
    protected final OnPermissionDeniedCallback deniedCallback;
    protected final OnPermissionShowRationaleCallback showRationaleCallback;

    public PermissionRequest(@NonNull PermissionManager permissionManager, @NonNull String[] permissions, int requestCode, OnPermissionGrantedCallback grantedCallback, OnPermissionDeniedCallback deniedCallback, OnPermissionShowRationaleCallback showRationaleCallback) {
        this.permissionManager = permissionManager;
        this.permissions = permissions;
        this.requestCode = requestCode;
        this.grantedCallback = grantedCallback;
        this.deniedCallback = deniedCallback;
        this.showRationaleCallback = showRationaleCallback;
    }

    public void acceptPermissionRationale() {
        permissionManager.requestPermission(this);
    }

    @NonNull
    protected String[] getPermissions() {
        return permissions;
    }

    protected int getRequestCode() {
        return requestCode;
    }

    protected void fireOnPermissionGrantedCallback() {
        if (grantedCallback != null) {
            grantedCallback.onPermissionGranted();
        }
    }

    protected void fireOnPermissionDeniedCallback(boolean neverAskAgain) {
        if (deniedCallback != null) {
            deniedCallback.onPermissionDenied(neverAskAgain);
        }
    }

    protected void fireOnPermissionShowRationaleCallback() {
        if (showRationaleCallback != null) {
            showRationaleCallback.onPermissionShowRationale(this);
        }
    }
}

File: library/src/main/java/com/hextremelabs/permiscus/PermissionRequest.kt
package com.hextremelabs.permiscus

import com.hextremelabs.permiscus.callbacks.OnPermissionDeniedCallback
import com.hextremelabs.permiscus.callbacks.OnPermissionGrantedCallback
import com.hextremelabs.permiscus.callbacks.OnPermissionShowRationaleCallback

class PermissionRequest(
    private val permissionManager: PermissionManager,
    val permissions: Array<String>,
    val requestCode: Int,
    private val grantedCallback: OnPermissionGrantedCallback?,
    private val deniedCallback: OnPermissionDeniedCallback?,
    private val showRationaleCallback: OnPermissionShowRationaleCallback?
) {
    fun acceptPermissionRationale() {
        permissionManager.requestPermission(this)
    }

    internal fun fireOnPermissionGrantedCallback() {
        grantedCallback?.onPermissionGranted()
    }

    internal fun fireOnPermissionDeniedCallback(neverAskAgain: Boolean) {
        deniedCallback?.onPermissionDenied(neverAskAgain)
    }

    internal fun fireOnPermissionShowRationaleCallback() {
        showRationaleCallback?.onPermissionShowRationale(this)
    }
}
